util/grpcutil: Can no longer detect unambiguous failures

SHA: https://github.com/cockroachdb/cockroach/commits/98f1623d60254505b65c457728617af5c5fec33b

Parameters:
```
TAGS=
GOFLAGS=
```

Stress build found a failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=394944&tab=buildLog

```
I171101 07:09:42.489844 41 vendor/google.golang.org/grpc/resolver_conn_wrapper.go:64  dialing to target with scheme: ""
I171101 07:09:42.489870 41 vendor/google.golang.org/grpc/resolver_conn_wrapper.go:70  could not get resolver for scheme: ""
W171101 07:09:42.490535 56 vendor/google.golang.org/grpc/server.go:839  grpc: Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"
W171101 07:09:42.490640 21 vendor/google.golang.org/grpc/transport/log.go:36  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:49561->127.0.0.1:33576: use of closed network connection
W171101 07:09:42.491825 52 vendor/google.golang.org/grpc/clientconn.go:934  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:49561: operation was canceled"; Reconnecting to {127.0.0.1:49561 0  <nil>}
W171101 07:09:42.491862 52 vendor/google.golang.org/grpc/clientconn.go:1028  grpc: addrConn.transportMonitor exits due to: context canceled
	grpc_util_test.go:117: request should not have started, but got rpc error: code = Unavailable desc = transport is closing


ERROR: exit status 1

40000 runs completed, 1 failures, over 4m25s
context canceled
Makefile:276: recipe for target 'stress' failed
make: *** [stress] Error 1
```

--- a/pkg/kv/transport.go
+++ b/pkg/kv/transport.go
@@ -248,6 +248,9 @@ func (gt *grpcTransport) send(
 		if err != nil {
 			return nil, err
 		}
+		if err := grpcutil.ConnectionReady(conn); err != nil {
+			return nil, err
+		}
 		reply, err := roachpb.NewInternalClient(conn).Batch(ctx, &client.args)
 		if reply != nil {
 			for i := range reply.Responses {

--- a/pkg/util/grpcutil/grpc_util.go
+++ b/pkg/util/grpcutil/grpc_util.go
@@ -16,6 +16,7 @@ package grpcutil
 
 import (
 	"context"
+	"fmt"
 	"io"
 	"strings"
 
@@ -24,6 +25,7 @@ import (
 
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/connectivity"
 	"google.golang.org/grpc/status"
 	"google.golang.org/grpc/transport"
 )
@@ -73,6 +75,9 @@ func IsClosedConnection(err error) bool {
 // TODO(bdarnell): Replace this with a cleaner mechanism when/if
 // https://github.com/grpc/grpc-go/issues/1443 is resolved.
 func RequestDidNotStart(err error) bool {
+	if _, ok := err.(connectionNotReadyError); ok {
+		return true
+	}
 	s, ok := status.FromError(err)
 	if !ok {
 		// This is a non-gRPC error; assume nothing.
@@ -86,4 +91,30 @@ func RequestDidNotStart(err error) bool {
 		return true
 	}
 	return false
+}
+
+// ConnectionReady returns nil if the given connection is ready to
+// send a request, or an error (which will pass RequestDidNotStart) if
+// not.
+//
+// This is a workaround for the fact that gRPC 1.7 fails to
+// distinguish between ambiguous and unambiguous errors.
+//
+// This is designed for use with connections prepared by
+// pkg/rpc.Connection.Connect (which performs an initial heartbeat and
+// thereby ensures that we will never see a connection in the
+// first-time Connecting state).
+func ConnectionReady(conn *grpc.ClientConn) error {
+	if s := conn.GetState(); s == connectivity.TransientFailure {
+		return connectionNotReadyError{s}
+	}
+	return nil
+}
+
+type connectionNotReadyError struct {
+	state connectivity.State
+}
+
+func (e connectionNotReadyError) Error() string {
+	return fmt.Sprintf("connection not ready: %s", e.state)
 }

